### ISSUE TITLE
How to extract all false positive and false negative images of test dataset

### ISSUE THREAD

**Issue by amankumarjain** [ultralytics/ultralytics#3358]
![confusion_matrix](https://github.com/ultralytics/ultralytics/assets/11255032/3808b957-86c1-4c3e-92be-a1d1493a9493)
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have executed `yolo val` command and get the output precision, recall n map for each class on the test dataset. Now I want to extract all frames which are false positives and false negatives.
ac to confusion matrix, i want to extract all

{class} as background images (129 images)
background as {class} images (53 images)
class = head

is there any way to extract these images after `yolo val model={model} data={data} split=test` command on the test dataset?

**Comment by glenn-jocher** [comment]
@amankumarjain hello,

Thank you for reaching out. To extract the false positive and false negative images from the test dataset after running the `yolo val` command, you can use the `--save-conf` flag.

After the validation process, you will have a `results` folder containing the validation results. Inside the `results` folder, you will find a `confusion_matrix.txt` file that provides the predicted and ground truth labels for each image.

To extract the false positive images, you can filter the images where the predicted label is the class you are interested in (e.g., background) and the ground truth label is a different class (e.g., head). Similarly, you can extract the false negative images by filtering the images where the predicted label is a different class (e.g., head) and the ground truth label is the class you are interested in (e.g., background).

You can then move or copy these images to a separate folder for further analysis or processing.

I hope this helps! Let me know if you have any further questions.

Best regards,
Glenn Jocher

**Comment by amankumarjain** [comment]
Hi @glenn-jocher 

Thanks for your response, I have tried using the flag but looks like it's not saving any results in folder
I have executed thus command `yolo val model=/home/vogo/Documents/nelson/imgPrc/yolov8/yolov8_t5/v0/result_nohub/t15/weights/best.pt data=data_t5.yaml split=test save_conf=True save_txt=True save=True`

I have attached my terminal screenshot, after running command only difference id there are one lables folder inside runs/detect/val6

am i missing something ?
<img width="1428" alt="Screenshot 2023-06-26 at 1 07 22 AM" src="https://github.com/ultralytics/ultralytics/assets/11255032/76dcc944-f014-4d48-b708-d7b96c7b4516">

**Comment by amankumarjain** [comment]
> Hi @glenn-jocher
> 
> Thanks for your response, I have tried using the flag but looks like it's not saving any results in folder I have executed thus command `yolo val model=/home/vogo/Documents/nelson/imgPrc/yolov8/yolov8_t5/v0/result_nohub/t15/weights/best.pt data=data_t5.yaml split=test save_conf=True save_txt=True save=True`
> 
> I have attached my terminal screenshot, after running command only difference id there are one lables folder inside runs/detect/val6
> 
> am i missing something ? <img alt="Screenshot 2023-06-26 at 1 07 22 AM" width="1428" src="https://user-images.githubusercontent.com/11255032/248583306-76dcc944-f014-4d48-b708-d7b96c7b4516.png">

@glenn-jocher please let me know if it's correct or not.

**Comment by glenn-jocher** [comment]
Hello @amankumarjain, thank you for getting back to me with the details. 

Regarding your issue with the `save_conf` flag not working properly, there seems to be a typo in the command you provided. The correct syntax for the flag is `--save-conf`, with two dashes instead of one. 

Therefore, the command should be `yolo val model=/home/vogo/Documents/nelson/imgPrc/yolov8/yolov8_t5/v0/result_nohub/t15/weights/best.pt data=data_t5.yaml split=test --save_conf=True save_txt=True save=True`. Please try running this command with the correct syntax and let me know if it resolves your issue.

If you continue to experience issues with the `--save-conf` flag, please provide me with more details on the version of YOLOv8 you are running and any error messages you are receiving.

I hope this helps and please let me know if you have any other questions or concerns.

Best regards,
Glenn Jocher.

**Comment by darouwan** [comment]
So can I use save-conf in python method also?

**Comment by glenn-jocher** [comment]
@darouwan hello,

Yes, you can use `save_conf` in the Python method as well. The `save_conf` argument allows you to save the confusion matrix results during the validation process.

To use `save_conf` in the Python method, you need to set it to `True` when calling the `YOLOv8().val()` function. This will save the confusion matrix results in the specified output folder.

For example:

```python
from yolov8 import YOLOv8

model = YOLOv8()
model.val(data='./data.yaml', save_conf=True, save_txt=True, save=True)
```

By setting `save_conf=True`, the confusion matrix will be saved along with other validation results.

I hope this clarifies your question. If you have any further inquiries, please let me know.

Best regards,
Glenn Jocher

**Comment by darouwan** [comment]
I validate my dataset again with the code you give but cannot get the `results `folder and `confusion_matrix.txt`.  Did I do something wrong? 

```
model = YOLO("myproject/train14/weights/best.pt")
model.val(data=r"F:\python_project\datasets\myproject\classes.yaml",
                 project='myproject', batch=64,save_conf=True, save_txt=True, save=True)
```

**Comment by glenn-jocher** [comment]
@darouwan hello,

Thank you for reaching out. 

In your code, it seems that you are using the correct syntax to save the confusion matrix results by setting `save_conf=True`. However, the `results` folder and `confusion_matrix.txt` file are not being generated, and you are wondering if you did something wrong.

There could be a few potential reasons for this issue. Here are some suggestions to troubleshoot:

1. Make sure that the `project` parameter provided to the `YOLO().val()` function matches the project name where your validation results should be saved.

2. Verify that you have the necessary write permissions for the directory where the `results` folder and `confusion_matrix.txt` file should be saved. If you don't have the proper permissions, the results may not be saved correctly.

3. Double-check that the `data` argument points to the correct YAML file containing your dataset information. Ensure that the file path is correct and that the YAML file is properly formatted.

If none of these suggestions resolve the issue, please provide more details, such as the version of YOLOv8 you are using, any error messages you are encountering, and any additional relevant code snippets. This will help in diagnosing the problem more accurately.

I hope this guidance helps you identify the issue. If you have any further questions or concerns, please don't hesitate to ask.

Best regards,
Glenn Jocher

**Comment by derek-knox** [comment]
I am not seeing the `results` folder or `confusion_matrix.txt` either. As of what version is this expected to work?

**Comment by darouwan** [comment]
> I am not seeing the `results` folder or `confusion_matrix.txt` either. As of what version is this expected to work?
> 
> Additionally, `save_conf` is not currently a listed argument [in the docs](https://docs.ultralytics.com/modes/val/#arguments)

My fork with dev branch can implement this feature, you can try it

**Comment by derek-knox** [comment]
Thanks @glenn-jocher. Your [previous comment](https://github.com/ultralytics/ultralytics/issues/3358#issuecomment-1605283807) contradicted this which is what led to the confusion.

**Comment by glenn-jocher** [comment]
@derek-knox apologies for the confusion in my previous message. That was a mistake on my part. Currently, in YOLOv8, there isn't a `save_conf` feature in the `val()` function available to save confusion matrix results to a `results` folder or `confusion_matrix.txt` file. 

If you're looking to implement this feature, you would need to write a custom function that takes the model's predictions, compares them with the ground truth labels, and generates a confusion matrix accordingly. 

Once again, I apologize for the incorrect information in my previous comment. Please feel free to reach out if you have more questions or need further assistance with YOLOv8.

**Comment by thaihuyends** [comment]
Hello @darouwan 
I use your repo to show print and save false positive and true negatives images, but I do not get it!
Here is my code, can you tell me where I am wrong? Thanks so much for your support!!

from ultralytics import YOLO

# Load a model
model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)

# Use the model
metrics = model.val(data='/Users/apple/Documents/ultralytics/truck/data.yaml', save_dir='results_val')

**Comment by darouwan** [comment]
@thaihuyends  Did you check the 'dev' branch? Only the dev branch has this feature.

**Comment by ddracoo** [comment]
> @thaihuyends Did you check the 'dev' branch? Only the dev branch has this feature.

so the dev branch has a save_conf argument inside the val mode? And how does it save the confidence / images?
Thanks!

**Comment by darouwan** [comment]
@ddracoo No need to set save_conf . The dev branch will output false positive and false negative images by default. The images are from confusion matrix.

**Comment by Vikas-ABD** [comment]
can anyone suggest me any repo to separate false positive and false negative images for each classes

**Comment by glenn-jocher** [comment]
@Vikas-ABD the Ultralytics YOLOv8 repository doesn't natively support extracting false positive and false negatives images for each class after validation out-of-the-box. Custom post-processing scripts are typically required to sort through validation results and separate these images based on the desired criteria. 

Always ensure your scripts are compatible with the structure of the validation output provided by YOLOv8. If a feature like this is highly requested, we might consider adding it in future updates. Stay tuned on the repo for any announcements. 🚀

**Comment by Vikas-ABD** [comment]
This is my solution for each classes at a time extracting FP and FN  and TP

GITHUB Link:https://github.com/Vikas-ABD/yolov8_custom_data_evaluation_56Secure_Assignment.git

**Comment by glenn-jocher** [comment]
@Vikas-ABD, it's great to see you exploring custom solutions for evaluating YOLOv8 models. Thanks for sharing your work with the community! 👍 If anyone is looking for a custom approach to extract false positives, false negatives, and true positives for each class for detailed analysis, they might find your GitHub repository helpful.

Just a reminder, always make sure that any third-party solutions you adopt or modify comply with the licensing agreement of the original codebase. Happy coding!

**Comment by clwehner** [comment]
@darouwan Thanks for your support in this case. 

I can not find any 'dev' branch on neither the ultralytics/ultralytics nor your darouwan/ultralytics repo. 
Since there are different ones available, can someone give me a hint which branch implements the save_conf feature to extract FP and FN?

Thx! :)

**Comment by darouwan** [comment]
> @darouwan Thanks for your support in this case. 
> 
> I can not find any 'dev' branch on neither the ultralytics/ultralytics nor your darouwan/ultralytics repo. 
> Since there are different ones available, can someone give me a hint which branch implements the save_conf feature to extract FP and FN?
> 
> Thx! :)

You can try my_feature branch for instead.
